### PR TITLE
Provide backoff for retryable aws requests and the option to adjust the cf status check interval via an environment variable

### LINKED ISSUE #6980: Adding a jitter to the backoff when an aws request gets throttled.
# Feature Proposal

Add exponential backoff instead of always 5 seconds.

## Description

We are deploying quite often to our test AWS account. In that I noticed that we are hitting cloudformation api throttling 

```
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Validating template...
Serverless: Creating Stack...
Serverless: Checking Stack create progress...
...Serverless: Recoverable error occurred (Rate exceeded), sleeping for 5 seconds. Try 1 of 4
Serverless: Recoverable error occurred (Rate exceeded), sleeping for 5 seconds. Try 1 of 4
Serverless: Recoverable error occurred (Rate exceeded), sleeping for 5 seconds. Try 2 of 4
Serverless: Recoverable error occurred (Rate exceeded), sleeping for 5 seconds. Try 3 of 4
Serverless: Recoverable error occurred (Rate exceeded), sleeping for 5 seconds. Try 4 of 4
```

[https://github.com/serverless/serverless/blob/801539ab2c1c94c5fc6d1c38c648df3fe5aec005/lib/plugins/aws/provider/awsProvider.js#L255](https://github.com/serverless/serverless/blob/801539ab2c1c94c5fc6d1c38c648df3fe5aec005/lib/plugins/aws/provider/awsProvider.js#L255)

This is at least what I am expecting that is happending. The issue is that all my stacks will now retrying in exactly 5 seconds causing them eventually to fail.

I think there should be some kind of jitter added in order to have a higher chance of success.

### PR DESCRIPTION
**Pull request by philiiiiiipp**
## What did you implement

Closes #6980 

I did create that inside the github console, given the simple logic I do believe that its fine. But those might also be famous last words!

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/provider/awsProvider.js [hunk 1]
```diff
@@ -243,16 +243,21 @@ class AwsProvider {
                 e.statusCode !== 403 && // Invalid credentials
                 ((e.providerError && e.providerError.retryable) || e.statusCode === 429)
               ) {
+                const nextTryNum = numTry + 1;
+                const exponentialBackOffTime = Math.round(
+                  5000 + nextTryNum * Math.random() * 10000
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js]: Maybe we can simplify, and set retry timeout to something between 4 and 7 seconds.

I wouldn't also add back off, it doesn't seem justified (we do not wait for long going request to finish).

File: lib/plugins/aws/provider/awsProvider.js [hunk 2]
```diff
@@ -243,16 +243,21 @@ class AwsProvider {
...
+                );
+
                 that.serverless.cli.log(
                   _.join(
                     [
-                      `Recoverable error occurred (${e.message}), sleeping for 5 seconds.`,
-                      `Try ${numTry + 1} of ${MAX_TRIES}`,
+                      `Recoverable error occurred (${e.message}), sleeping for ${exponentialBackOffTime / 1000} seconds.`,
```
Comment (medikoo) [inline lib/plugins/aws/provider/awsProvider.js]: Maybe let's change it to `sleeping for few seconds`

Comment (philiiiiiipp) [reply lib/plugins/aws/provider/awsProvider.js]: We could do that, but I would prefer to have the exact number when looking at the logs. Especially now that we would move away from 5 seconds to a variable amount.

If you do not have a strong opinion about it I would leave it as is.

Comment (medikoo) [reply lib/plugins/aws/provider/awsProvider.js]: We will not have fixed seconds number.. it'll be e.g. 4.234s, so it's a bit noisy.

Why do you find this output valuable? Specifically to confirm that jitter functionality works?

File: lib/plugins/aws/provider/awsProvider.js (hunk 1 above)
Comment (medikoo) [reply lib/plugins/aws/provider/awsProvider.js]: @philiiiiiipp have you seen this comment maybe?

**Comment by philiiiiiipp** [comment]
I somehow cannot answer on your comments anymore. I believe that last unresolved one was about the backoff. 

In my opinion its a good pattern to have, but having a jitter in combination with the ability to increase the interval at the other place would suffice for my use case at least. I am not that knowledgeable in the codebase though, so if you prefer not having an increasing backoff, I am cool with that too.

**Comment by medikoo** [comment]
> In my opinion its a good pattern to have

Can you elaborate why you see it useful in that case?

To me backoff serves well cases where we deal with a long wait time, and there's no point to check often. While here we deal with cases where there are short requests, and issue is that in some cases they're issued at very same time, so with fixed frequency they clash. In such case jitter seems the only solution that could indeed help,

I would like to avoid adding functionalities that do not serve real purpose (do not complicate things when it's not justified)

**Comment by philiiiiiipp** [comment]
Right, we are doing requests against an endpoint, which might be congested, not only because of sls, but in general because there are many other clients trying to reach it. So if you want to allow the endpoint to recover it makes sense to slowly back off until your retries are used up.

https://docs.aws.amazon.com/general/latest/gr/api-retries.html

Like I said I am not married to this, I am more eager to get the jitter and the custom frequency option in, in order to stop using our forked version. So if you feel that this makes more harm than good I will remove it and we can always revisit it at a later stage.

**Comment by medikoo** [comment]
> So if you feel that this makes more harm than good I will remove it

My feeling is that adding extra 5s every time, in this specific case, will rather bring experience downgrade than upgrade.

I would also narrow wait range to something as 4-7s.  The way it's configured now is 5-10s, so again that will make waits noticeably longer from what we have in master.

Technically I think it'll be good to focus in this PR purely on adding jitter, and maintain interval time on similar level as we have now.

The other reason I'm reluctant to make any bigger changes here, is that I don't remember having any other reports, that concerns that retry logic. So ideal approach seems to fix it for you by adding a jitter, but at same time maintain similar interval times, to keep experience not different for all other framework users.

**Comment by philiiiiiipp** [comment]
> I would also narrow wait range to something as 4-7s. The way it's configured now is 5-10s, so again that will make waits noticeably longer from what we have in master.
> 
> Technically I think it'll be good to focus in this PR purely on adding jitter, and maintain interval time on similar level as we have now.
> 
> The other reason I'm reluctant to make any bigger changes here, is that I don't remember having any other reports, that concerns that retry logic. So ideal approach seems to fix it for you by adding a jitter, but at same time maintain similar interval times, to keep experience not different for all other framework users.

Right, that makes sense, I adjusted the interval to 4-7 seconds and removed the increasing backoff.

**Pull request merged**